### ISSUE TITLE
Better parameter value access

### ISSUE THREAD

**Issue by hdoupe** [PSLmodels/ParamTools#12]
ParamTools does not make it easy to retrieve the value of a given parameter. There are a couple methods that can be used to retrieve values, but they have their limitations:

- `Params.get`: This is to verbose. It is annoying to have to specify the queried dimensions if they do not change frequently.
```python
In [2]: params.get("average_precipitation", month="November", city="Washington, D.C.")
Out[2]: [{'month': 'November', 'city': 'Washington, D.C.', 'value': 3.0}]

In [3]: params.get("average_precipitation", month="November", city="Washington, D.C.")[0]["value"]
Out[3]: 3.0

```

- `Params.to_array`: The user can not query along different dimensions and only dense value lists are supported (i.e. each value in the parameter's the parameter space must be specified). 
```python
In [4]: params.to_array("average_precipitation")
Out[4]: 
array([[3.1, 2.6, 3.5, 3.3, 4.3, 4.3, 4.6, 3.8, 3.9, 3.7, 3. , 3.5],
       [3.6, 3.7, 4.3, 3.5, 3.8, 3.6, 5. , 3.8, 3.7, 2.8, 3.6, 4.1]])
```

`Parameters` is going to have to store some kind of state to solve this problem. By storing state, the number of values that will be returned can be reduced. For example, you could set the state for `month` to `November` and then only two values would be returned:

```python
params.set_state(month="November")
params.get("average_precipitation")

# output:
[{'month': 'November', 'city': 'Washington, D.C.', 'value': 3.0},
 {'month': 'November', 'city': 'Atlanta, GA', 'value': 3.6}]

params.to_array("average_precipitation")

#output: 
[[3.0], [3.6]]

```

It seems like a combination of state storage and a beefing up of the `to/from_array` methods will go a long ways towards solving the parameter value access problem.

**Comment by hdoupe** [comment]
ParamTools' goal is for projects to have the *choice* to access parameter values as arrays by default. Array-first access means that when accessing a parameter's value through the `Parameters` instance attribute (e.g. `Parameters.some_param_name`) the result should be an array of parameter values. Using a [modified version](https://github.com/PSLmodels/ParamTools/commit/c630ad2b8dd23c36c0c863910220920c9e7d81c0) of the taxcalc data example, array-first access would look like this:

```python
In [2]: from marshmallow import fields, Schema
   ...: 
   ...: from paramtools import Parameters
   ...: from paramtools.utils import get_example_paths
   ...: 
   ...: class CompatibleDataSchema(Schema):
   ...:     """
   ...:     Schema for Compatible data object
   ...:     {
   ...:         "compatible_data": {"data1": bool, "data2": bool, ...}
   ...:     }
   ...:     """
   ...: 
   ...:     puf = fields.Boolean()
   ...:     cps = fields.Boolean()
   ...: 
   ...: schema_path, defaults_path = get_example_paths("taxcalc")
   ...:
   ...: class TaxcalcParams(Parameters):
   ...:     schema = schema_path
   ...:     defaults = defaults_path
   ...:     field_map = {"compatible_data": fields.Nested(CompatibleDataSchema())}
   ...:     array_first = True
   ...: 
   ...: params = TaxcalcParams()
   ...: 
   ...: 

In [3]: params
Out[3]: <__main__.TaxcalcParams at 0x1054a9128>

In [4]: params._STD
Out[4]: 
array([[ 6100.  , 12200.  ,  6100.  ,  8950.  , 12200.  ],
       [ 6200.  , 12400.  ,  6200.  ,  9100.  , 12400.  ],
       [ 6300.  , 12600.  ,  6300.  ,  9250.  , 12600.  ],
       [ 6300.  , 12600.  ,  6300.  ,  9300.  , 12600.  ],
       [ 6350.  , 12700.  ,  6350.  ,  9350.  , 12700.  ],
       [12000.  , 24000.  , 12000.  , 18000.  , 24000.  ],
       [12238.8 , 24477.6 , 12238.8 , 18358.2 , 24477.6 ],
       [12490.92, 24981.84, 12490.92, 18736.38, 24981.84],
       [12756.98, 25513.95, 12756.98, 19135.46, 25513.95],
       [13035.08, 26070.15, 13035.08, 19552.61, 26070.15],
       [13311.42, 26622.84, 13311.42, 19967.13, 26622.84],
       [13597.62, 27195.23, 13597.62, 20396.42, 27195.23],
       [13887.25, 27774.49, 13887.25, 20830.86, 27774.49],
       [ 7655.  , 15311.  ,  7655.  , 11272.  , 15311.  ]])
```

For comparison, this is how `taxcalc` does parameter access:
```python
In [5]: from taxcalc import Policy

In [6]: pol = Policy()

In [7]: pol._STD
Out[7]: 
array([[ 6100.  , 12200.  ,  6100.  ,  8950.  , 12200.  ],
       [ 6200.  , 12400.  ,  6200.  ,  9100.  , 12400.  ],
       [ 6300.  , 12600.  ,  6300.  ,  9250.  , 12600.  ],
       [ 6300.  , 12600.  ,  6300.  ,  9300.  , 12600.  ],
       [ 6350.  , 12700.  ,  6350.  ,  9350.  , 12700.  ],
       [12000.  , 24000.  , 12000.  , 18000.  , 24000.  ],
       [12238.8 , 24477.6 , 12238.8 , 18358.2 , 24477.6 ],
       [12490.92, 24981.84, 12490.92, 18736.38, 24981.84],
       [12756.98, 25513.95, 12756.98, 19135.46, 25513.95],
       [13035.08, 26070.15, 13035.08, 19552.61, 26070.15],
       [13311.42, 26622.84, 13311.42, 19967.13, 26622.84],
       [13597.62, 27195.23, 13597.62, 20396.42, 27195.23],
       [13887.25, 27774.49, 13887.25, 20830.86, 27774.49],
       [ 7655.  , 15311.  ,  7655.  , 11272.  , 15311.  ],
       [ 7818.82, 15638.66,  7818.82, 11513.22, 15638.66]])

In [8]: 
```

Note that ParamTools can also set the state of the `Parameters` instance:

```python
In [8]: params.set_state(year=2019)

In [9]: params._STD
Out[9]: array([[12238.8, 24477.6, 12238.8, 18358.2, 24477.6]])

In [10]: params.clear_state() # go back to empty state

In [11]: params._STD
Out[11]: 
array([[ 6100.  , 12200.  ,  6100.  ,  8950.  , 12200.  ],
       [ 6200.  , 12400.  ,  6200.  ,  9100.  , 12400.  ],
       [ 6300.  , 12600.  ,  6300.  ,  9250.  , 12600.  ],
       [ 6300.  , 12600.  ,  6300.  ,  9300.  , 12600.  ],
       [ 6350.  , 12700.  ,  6350.  ,  9350.  , 12700.  ],
       [12000.  , 24000.  , 12000.  , 18000.  , 24000.  ],
       [12238.8 , 24477.6 , 12238.8 , 18358.2 , 24477.6 ],
       [12490.92, 24981.84, 12490.92, 18736.38, 24981.84],
       [12756.98, 25513.95, 12756.98, 19135.46, 25513.95],
       [13035.08, 26070.15, 13035.08, 19552.61, 26070.15],
       [13311.42, 26622.84, 13311.42, 19967.13, 26622.84],
       [13597.62, 27195.23, 13597.62, 20396.42, 27195.23],
       [13887.25, 27774.49, 13887.25, 20830.86, 27774.49],
       [ 7655.  , 15311.  ,  7655.  , 11272.  , 15311.  ]])

```

**Comment by hdoupe** [comment]
The value access issue (#12) was resolved via #22, #23, and #24. I'm eager to hear feedback on this solution and how it can be improved.

cc @MattHJensen @jdebacker @codykallen
